language/move-vm/runtime/src/loader.rs

```diff
@@ -0,0 +1,1378 @@
+struct ScriptCache {
+    scripts: BinaryCache<HashValue, Script>,
+}
+
+impl ScriptCache {
+    fn new() -> Self {
+        Self {
+            scripts: BinaryCache::new(),
+        }
+    }
+
+    fn get(&self, hash: &HashValue) -> Option<Arc<Function>> {
+        self.scripts.get(hash).map(|script| script.entry_point())
+    }
+
+    fn insert(&mut self, hash: HashValue, script: Script) -> VMResult<Arc<Function>> {
+        match self.get(&hash) {
+            Some(script) => Ok(script),
+            None => Ok(self.scripts.insert(hash, script).entry_point()),
+        }
+    }
+}
+
+// A ModuleCache is the core structure in the Loader.
+// It holds all Modules, Types and Functions loaded.
+// Types and Functions are pushed globally to the ModuleCache.
+// A ModuleCache is accessed under lock.
+struct ModuleCache {
+    modules: BinaryCache<ModuleId, Module>,
+    structs: Vec<Arc<StructType>>,
+    functions: Vec<Arc<Function>>,
+}
+
+impl ModuleCache {
+    fn new() -> Self {
+        Self {
+            modules: BinaryCache::new(),
+            structs: vec![],
+            functions: vec![],
+        }
+    }
+
+    fn get(&self, id: &ModuleId) -> Option<Arc<Module>> {
+        self.modules.get(id).map(|module| Arc::clone(module))
+    }
+
+    fn insert(&mut self, id: ModuleId, ver_module: VerifiedModule) -> VMResult<Arc<Module>> {
+        self.publish_module(&ver_module)?;
+        let module = Module::new(ver_module, self)?;
+        match self.get(&id) {
+            Some(module) => Ok(module),
+            None => Ok(Arc::clone(self.modules.insert(id, module))),
+        }
+    }
+
+    fn function_at(&self, idx: usize) -> Arc<Function> {
+        self.functions[idx].clone()
+    }
+
+    fn struct_at(&self, idx: usize) -> Arc<StructType> {
+        Arc::clone(&self.structs[idx])
+    }
+
+    fn publish_module(&mut self, module: &VerifiedModule) -> VMResult<()> {
+        let starting_idx = self.structs.len();
+        for (idx, struct_def) in module.struct_defs().iter().enumerate() {
+            let st = self.load_type(module, struct_def, StructDefinitionIndex(idx as u16))?;
+            self.structs.push(Arc::new(st));
+        }
+        self.load_fields(module, starting_idx)?;
+
+        for func in module.function_defs() {
+            let function = self.load_function(module, func)?;
+            self.functions.push(Arc::new(function));
+        }
+        Ok(())
+    }
+
+    fn load_type(
+        &self,
+        module: &VerifiedModule,
+        struct_def: &StructDefinition,
+        idx: StructDefinitionIndex,
+    ) -> VMResult<StructType> {
+        let struct_handle = module.struct_handle_at(struct_def.struct_handle);
+        let is_resource = struct_handle.is_nominal_resource;
+        let name = module.identifier_at(struct_handle.name).to_owned();
+        let type_parameters = struct_handle.type_parameters.clone();
+        let module = module.self_id();
+        Ok(StructType {
+            fields: vec![],
```

Comment:
Any reason we're loading the struct type and fields separately? I got confused by the starting_idx below initially.

Reply:
we cannot load fields at the same time as types because the types have not been "published" yet and thus we don't have the indexes for the type of the fields, so that must be done in 2 steps